Gems/ROS2/Code/Source/RobotImporter/Utils/RobotImporterUtils.cpp

```diff
@@ -90,132 +89,294 @@ namespace ROS2
+
+        sdf::SemanticPose linkSemanticPos = link->SemanticPose();
+        gz::math::Pose3d resolvedPose;
+
+        if (sdf::Errors poseResolveErrors = linkSemanticPos.Resolve(resolvedPose);
+            !poseResolveErrors.empty())
         {
-            t = URDF::TypeConversions::ConvertPose(link->parent_joint->parent_to_joint_origin_transform) * t;
-            return GetWorldTransformURDF(link->getParent(), t);
+            AZStd::string poseErrorMessages;
+            for (const sdf::Error& error : poseResolveErrors)
+            {
+                AZStd::string errorMessage = AZStd::string::format("ErrorCode=%d", static_cast<int32_t>(error.Code()));
+                errorMessage += AZStd::string::format(", Message=%s", error.Message().c_str());
+                if (error.LineNumber().has_value())
+                {
+                    errorMessage += AZStd::string::format(", Line=%d", error.LineNumber().value());
+                }
+
+                poseErrorMessages += errorMessage;
+                poseErrorMessages += '\n';
+            }
+
+            AZ_Error("RobotImporter", false, R"(Failed to get world transform for link %s. Errors: "%s")",
+                link->Name().c_str(), poseErrorMessages.c_str());
+            return {};
         }
-        return t;
+
+        const AZ::Transform linkTransform = URDF::TypeConversions::ConvertPose(resolvedPose);
+        const AZ::Transform resolvedTransform = linkTransform * t;
+        return resolvedTransform;
+    }
+
+    void VisitLinks(const sdf::Model& sdfModel, const LinkVisitorCallback& linkVisitorCB,
+        bool visitNestedModelLinks)
+    {
+        // Function object which can visit all links of a model
+        // Optionally it supports recursing nested models to visit their links as well
+        struct VisitLinksForNestedModels_fn
+        {
+            void operator()(const sdf::Model& model)
+            {
+                VisitLinksForModel(model);
+                if (m_recurseModels)
+                {
+                    for (uint64_t modelIndex{}; modelIndex < model.ModelCount(); ++modelIndex)
+                    {
+                        const sdf::Model* nestedModel =  model.ModelByIndex(modelIndex);
+                        if (nestedModel != nullptr)
+                        {
+                            VisitLinksForModel(*nestedModel);
+                        }
+                    }
+                }
+            }
+
+        private:
+            void VisitLinksForModel(const sdf::Model& currentModel)
+            {
+                for (uint64_t linkIndex{}; linkIndex < currentModel.LinkCount(); ++linkIndex)
+                {
+                    const sdf::Link* link = currentModel.LinkByIndex(linkIndex);
+                    if (link != nullptr)
+                    {
+                        m_linkVisitorCB(*link);
```

Comment:
Suggestion: For visit patterns, it's often helpful for the callback to return a bool that determines whether or not to continue processing. This allows the callback to short-circuit on failures, matches, etc.

Reply:
I'll make that update.